File: Python/WINTER_PRACTICE/3_task/Gauss/main.cpp
#include <iostream>
#include <cmath>
#include <vector>
#include <iomanip>

using namespace std;

// ------------------ Gauss Quadrature for Double Integrals ------------------
double Gauss_Quadrature_2d(double (*f)(double, double), double a, double A, double b, double B, int n, int p) {
    vector<double> roots, weights;

    if (p == 5) {
        roots = {-0.906179845938664, -0.538469310105683, 0, 0.538469310105683, 0.906179845938664};
        weights = {0.236926885056189, 0.478628670499366, 0.568888888888889, 0.478628670499366, 0.236926885056189};
    } else if (p == 4) {
        roots = {-0.861136311594053, -0.339981043584856, 0.339981043584856, 0.861136311594053};
        weights = {0.3478548451374529, 0.6521451548625463, 0.6521451548625463, 0.3478548451374529};
    } else if (p == 3) {
        roots = {-0.774596669241483, 0, 0.774596669241483};
        weights = {0.555555555555556, 0.888888888888889, 0.555555555555556};
    } else if (p == 2) {
        roots = {-0.577350269189626, 0.577350269189626};
        weights = {1, 1};
    } else {
        roots = {0};
        weights = {2};
    }

    double h_a = (A - a) / n;
    double h_b = (B - b) / n;
    double integral = 0;

    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            double a_tmp = a + i * h_a;
            double A_tmp = a_tmp + h_a;
            double b_tmp = b + j * h_b;
            double B_tmp = b_tmp + h_b;

            double tmp_integral = 0;

            for (int k = 0; k < p; k++) {
                double x_mapped = (A_tmp - a_tmp) / 2 * roots[k] + (A_tmp + a_tmp) / 2;

                for (int l = 0; l < p; l++) {
                    double y_mapped = (B_tmp - b_tmp) / 2 * roots[l] + (B_tmp + b_tmp) / 2;
                    tmp_integral += weights[k] * weights[l] * f(x_mapped, y_mapped);
                }
            }

            integral += ((A_tmp - a_tmp) / 2) * ((B_tmp - b_tmp) / 2) * tmp_integral;
        }
    }

    return integral;
}


double func(double x, double y) { return 2 * x * sin(x * y); }

// ------------------ My Integral ------------------

int main(){
    double eps;

    cout << "Enter a floating-point number eps, ";
    cout << "that is, the precision with which you want to numerically find approximate value of the given integral" << endl;
    cin >> eps;

    double a = 0;
    double A = 1;

    double b = 0;
    double B = M_PI / 2;
    
    int n = 2;
    int p = n;
    int alg_presicion = 2 * n - 1;
    
    cout << "n = " << n << endl;
    
    double result_n = Gauss_Quadrature_2d(func, a, A, b, B, n, p);
    double result_2n = Gauss_Quadrature_2d(func, a, A, b, B, 2 * n, p);
    
    double E = abs(result_n - result_2n) / (pow(2, alg_presicion) - 1);
    cout << "Error: " << E << endl;
    
    // ------------------ Runge Accuracy Control ------------------
    while(E >= eps){
        n = 2 * n;
        result_n = result_2n;
        result_2n = Gauss_Quadrature_2d(func, a, A, b, B, 2 * n, p);
        E = abs(result_n - result_2n) / (pow(2, alg_presicion) - 1);
        cout << "n = " << n << endl;
        cout << "Error: " << E << endl;
    }
    
    double z = log(eps)/log(10);

    cout << "Trapezoidal rule (h): " << setprecision(z + 1) << result_n;
}

File: Python/WINTER_PRACTICE/3_task/Gauss/main.py
import numpy as np
import sympy as sp

# ------------------ Legendre Polynomial ------------------
def legendre_poly(n, x):
    P_0 = 1
    P_1 = x
    
    if n == 0:
        return P_0
    elif n == 1:
        return P_1
    
    P_prev2 = P_0
    P_prev1 = P_1
    P_current = P_0
    
    for i in range(2, int(n + 1)):
        P_current = ((2 * i - 1) * x * P_prev1 - (i - 1) * P_prev2) / i
        P_prev2 = P_prev1
        P_prev1 = P_current
    
    return P_current

# ------------------ Gauss Quadrature for Double Integrals ------------------
def Gauss_Quadrature_2d(f, a, A, b, B, n):
    x = sp.symbols('x')
    legendre_poly_expr = legendre_poly(n, x)
    roots = sp.solvers.solve(legendre_poly_expr, x)
    roots = [float(root) for root in roots]
    
    weights = []
    x_roots_transformed = []
    y_roots_transformed = []
    for root in roots:
        weight = 2 / ((1 - root**2) * (sp.diff(legendre_poly_expr, x).subs(x, root))**2)
        weights.append(float(weight))
        x_roots_transformed.append((A - a) / 2 * root + (A + a) / 2)
        y_roots_transformed.append((B - b) / 2 * root + (B + b) / 2)
    
    integral = 0
    for i in range(int(n)):
        for j in range(int(n)):
            integral += (weights[i] * weights[j] * f(x_roots_transformed[i], y_roots_transformed[j]))
    
    return ((A - a) / 2) * ((B - b) / 2) * integral

# ------------------ My Integral ------------------


eps = 0.001

f = lambda x, y: 2 * x * sp.sin(x * y)

a, A = 0, 1
b, B = 0, np.pi / 2

n = 2
alg_presicion = 2 * n - 1

print("n =", n)

result_h = Gauss_Quadrature_2d(f, a, A, b, B, n)
result_2h = Gauss_Quadrature_2d(f, a, A, b, B, 2 * n)

E = abs(result_h - result_2h) / (2**(alg_presicion) - 1)
print("Error:", E.evalf())

# ------------------ Runge Accuracy Control ------------------
while E >= eps:
    n = 2 * n
    result_h = Gauss_Quadrature_2d(f, a, A, b, B, n)
    result_2h = Gauss_Quadrature_2d(f, a, A, b, B, 2 * n)
    E = abs(result_h - result_2h) / (2**(alg_presicion) - 1)
    print("n =", n)
    print("Error:", E.evalf())

print("Gauss quadrature formula(h):", round(result_h.evalf(), 4))

x, y = sp.symbols('x y')
exact_result = sp.integrate(f(x, y), (x, a, A), (y, b, B))
print("Reference(sympy):", exact_result.evalf())